### PR TITLE
Refactor/fix type errors

### PR DESCRIPTION
**Pull request by areisle**
This PR is just fixing as many type errors as possible and adding types for the untyped v3 schema package to make the migration to schema v4 easier. This fixes 478 type errors, but there are still 398 remaining. Most of the remaining errors require modifying the logic to fix and so are being left for a future PR

### CODE REVIEW

File: src/views/GraphView/components/GraphComponent/GraphComponent.tsx [hunk 1]
```diff
@@ -238,7 +256,7 @@ function GraphComponent(props: GraphComponentProps) {
    * @param {object} graphObjects graphObjects attached to simulation
    * @param {object} pMap propsMap keeps track of node/link properties and associated values
    */
-  const addGraphObject = useCallback((type, graphData, prop1, prop2, graphObjs, pMap) => {
+  const addGraphObject = useCallback((type: 'link' | 'node', graphData: GeneralRecordType, prop1, prop2, graphObjs: InitialGraphDataType['graphObjects'], pMap: PropsMap) => {
```
Comment (kttkjl) [inline src/views/GraphView/components/GraphComponent/GraphComponent.tsx:259]: not a blocker, but these type definitions are getting quite long, it might be worth it to multi-line it

File: src/views/GraphView/components/GraphComponent/kbgraph.tsx [hunk 2]
```diff
@@ -250,8 +342,8 @@ class GraphOptions {
   /**
    * Returns the color of the given object, given the current color property.
    */
-  getColor(obj, type) {
-    const { [`${type}Color`]: targetColor, [`${type}Colors`]: ColorMap } = this;
+  getColor(obj: GraphNode | GraphLink, type: 'links' | 'nodes'): string | undefined {
+    const { [`${type}Color` as 'linksColor' | 'nodesColor']: targetColor, [`${type}Colors` as 'linksColors' | 'nodesColors']: ColorMap } = this;
```
Comment (kttkjl) [inline src/views/GraphView/components/GraphComponent/kbgraph.tsx:346]: likewise, this is pretty hard to follow

File: src/views/MainView/index.tsx [hunk 3]
```diff
@@ -66,8 +66,8 @@ const Main = () => {
               <Route component={ErrorView} exact path="/error" />
               <AuthenticatedRoute component={AboutView} path="/about" />
               <AuthenticatedRoute component={ActivityView} path="/activity" />
-              <AuthenticatedRoute component={QuickSearch} exact path="/query" />
-              <AuthenticatedRoute component={AdvancedSearchView} exact path="/query-advanced" />
+              <AuthenticatedRoute component={QuickSearch} path="/query" />
+              <AuthenticatedRoute component={AdvancedSearchView} path="/query-advanced" />
```
Comment (kttkjl) [inline src/views/MainView/index.tsx:70]: is there a benefit in NOT having it exact?

**Review by kttkjl — COMMENTED** [review]
Just some questions, and some styling, else checks out

File: src/components/hooks/useObject.tsx [hunk 4]
```diff
@@ -9,8 +9,8 @@ import useDeepCompareEffect from 'use-deep-compare-effect';
  *
  * @param {Object} initialValue the start object
  */
-const useObject = (initialValue = {}) => {
-  const [content, setContent] = useReducer((state, action) => {
+function useObject<T extends { [key: string]: any }>(initialValue?: T | null | undefined) {
```
[the remark this answers is not in the document]
Comment (areisle) [reply src/components/hooks/useObject.tsx:12]: they don't behave quite the same no, this is more permissive than using `Record`. if you try to use `Record` here you'll get type errors

File: src/views/ActivityView/index.tsx [hunk 5]
```diff
@@ -133,7 +133,6 @@ const ActivityView = () => {
               sortable: true,
             },
           ]}
-          data={recentRecords}
```
[the remark this answers is not in the document]
Comment (areisle) [reply src/views/ActivityView/index.tsx:136]: it's because above they're using the gridapi to set the row data so there's no reason to pass the data as well

File: src/views/GraphView/components/GraphComponent/GraphComponent.tsx (hunk 1 above)
Comment (areisle) [reply src/views/GraphView/components/GraphComponent/GraphComponent.tsx:259]: I was avoiding swapping to multi lines to keep the diff more readable. this repo has eslint line length rule turned off but maybe in a future PR that should be turned on and all the violations fixed

File: src/views/MainView/index.tsx (hunk 3 above)
Comment (areisle) [reply src/views/MainView/index.tsx:70]: `exact` was not a prop that exists on the `AuthenticatedRoute` component so this prop was doing nothing

File: src/components/FormLayout/__tests__/index.test.tsx [hunk 6]
```diff
@@ -42,24 +42,9 @@ describe('FormLayout', () => {
     );
 
     expect(getByText('The username')).toBeInTheDocument();
+    fireEvent.click(getByText('Expand to see all optional fields'));
     expect(getByText('@rid')).toBeInTheDocument();
     expect(getByTestId('@rid')).toBeInTheDocument();
     expect(getByTestId('@rid').value).toEqual('#3:4');
   });
-
-  test('exclusion works', () => {
```
[the remark this answers is not in the document]
Comment (areisle) [reply src/components/FormLayout/__tests__/index.test.tsx:50]: because this prop was unused and therefore always used the default so was turned into a constant so there's nothing to test

File: src/components/DetailDrawer/index.tsx [hunk 7]
```diff
@@ -201,7 +195,7 @@ function DetailDrawer(props: DetailDrawerProps) {
   const drawerIsOpen = Boolean(node);
```
[the remark this answers is not in the document]
Comment (areisle) [reply src/components/DetailDrawer/index.tsx:195]: its used on lines 300 & 305. the swap from `if(Boolean(node))` to `if (node)` is because of a typescript quirk https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgLIE8DyAjAVhBMZAbwChkLkALCAG1oHsAuZAZzClAHMBuUgX1KlaEImCrdkAXmQgArvWRxWaLHgJEAPrIW1k2uSAAmEGKAhG+pYDGQAKcdwCUJcpQQMQrBiIB0jLgcJEC5fGnoGJwEhG3sAIQYfCDgQIOcXMkpkDy8k-wZAxxCwukYowSA

File: src/components/FormLayout/FieldGroup.tsx [hunk 8]
```diff
@@ -1,35 +1,37 @@
...
+
 /**
  * returns an array of strings without any of the indicated exclusion values
  *
  * @param {Array.<string>} orderingList specifies property display ordering
  * @param {Array.<string>} exclusions fields that should not be rendered
  */
-const exclusionFilter = (orderingList, exclusionList) => {
-  const newOrdering = [];
+const exclusionFilter = (orderingList: (string | string[])[], exclusionList: string[]) => {
```
[the remark this answers is not in the document]
Comment (areisle) [reply src/components/FormLayout/FieldGroup.tsx]: [06ec462](https://github.com/bcgsc/pori_graphkb_client/pull/72/commits/06ec462eb8c19cb5203132db447e28d186321638)

File: src/views/AboutView/components/AboutClasses/components/ClassDescription.tsx [hunk 9]
```diff
@@ -26,7 +26,7 @@ const ClassDescription = ({ name, description }: ClassDescriptionProps) => {
       const [result] = await api.query(body);
       return result;
     },
-    { staleTime: Infinity, throwOnError: false },
```
[the remark this answers is not in the document]
Comment (areisle) [reply src/views/AboutView/components/AboutClasses/components/ClassDescription.tsx:29]: `throwOnError` isn't an option when using use query

**Review by kttkjl — APPROVED** [review]
